packages/bus-query-operation/lib/Bindings.ts

```diff
@@ -148,29 +153,93 @@ export function materializeOperation(
+      // This will result in non-projected variables being replaced with their InitialBindings values.
+      let subBindings = bindingsFactory.fromBindings(bindings);
+      for (const binding of bindings) {
+        for (const curVariable of op.variables) {
+          if (termToString(curVariable) === termToString(binding[0])) {
+            subBindings = subBindings.delete(binding[0]);
+            break;
+          }
+        }
+      }
+
+      // Find projected variables which are present in the InitialBindings
+      // This will result in projected variables being handled via a values clause.
+      const values: Algebra.Operation[] = [];
+      const overlappingVariables: RDF.Variable[] = [];
+      const overlappingBindings: Record<string, RDF.Literal | RDF.NamedNode>[] = [];
+      originalBindings = originalBindings ?? bindings;
+      for (const currentVariable of op.variables) {
+        if (originalBindings.has(currentVariable)) {
+          const newBinding = { [termToString(currentVariable)]:
+            <RDF.NamedNode | RDF.Literal> originalBindings.get(currentVariable) };
 
-      // Only include projected variables in the sub-bindings that will be passed down recursively.
-      // If we don't do this, we may be binding variables that may have the same label, but are not considered equal.
-      const subBindings = bindingsFactory.bindings(<[RDF.Variable, RDF.Term][]> op.variables.map((variable) => {
-        const binding = bindings.get(variable);
-        if (binding) {
-          return [ variable, binding ];
+          overlappingVariables.push(currentVariable);
+          overlappingBindings.push(newBinding);
+          values.push(factory.createValues([ currentVariable ], [ newBinding ]));
         }
-        // eslint-disable-next-line no-useless-return,array-callback-return
-        return;
-      }).filter(Boolean));
+      }
+
+      let recursionResult: Algebra.Operation = materializeOperation(
+        op.input,
+        subBindings,
+        bindingsFactory,
+        options,
+        originalBindings,
+      );
+
+      if (values.length > 0) {
+        recursionResult = factory.createJoin([ ...values, recursionResult ]);
+      }
 
       return {
         recurse: false,
-        result: factory.createProject(
-          materializeOperation(
-            op.input,
-            subBindings,
-            bindingsFactory,
-            options,
-          ),
-          variables,
+        result: factory.createProject(recursionResult, op.variables),
+      };
+    },
+    filter(op: Algebra.Filter, factory: Factory) {
+      originalBindings = originalBindings ?? bindings;
+
+      if (op.expression.expressionType !== 'operator' || originalBindings.size === 0) {
+        return {
+          recurse: false,
+          result: op,
+        };
+      }
+
+      // Make a values clause using all the variables from InitialBindings
+      const values: Algebra.Operation[] = [];
```

Comment:
I implemented this, unfortunately I had to iterate over bindings rather than variables in the outside loop, so I couldn't make use of bindings.get() anymore. It should be less efficient like this :/